lib/fs/fs.ml

```diff
@@ -67,7 +79,11 @@ let rec to_tree path =
     in
     let dirname = Filename.basename path in
     Dir (dirname, children)
-  else File (Filename.basename path, lazy (read_file_contents path))
+  else
+    File
+      ( Filename.basename path,
+        lazy (read_file_contents path),
+        lazy (read_file_type path) )
```

Comment:
Agree, reading the file is not ideal. In that case, let's do the following:

* Move `file_type` inside the `file_contents` record

This way the implementation almost won't change and we'll read the file only once.

<hr>

Strictly speaking, semantically, instead of having a separate enum, the `file_contents` type actually should change to:

```ocaml
type file_contents =
  | Binary
  | Text of { lines : Pretty.doc array; offset : int }
```

This data type reflects the actual state of things better and the actual way to display the file is now handled by functions under `view`.

I'm not sure if it'll make the `file_cursor` logic more complex. Maybe the changes won't be that big.

So feel free to try that approach too if you feel it's better!